# Data Exploration and Cleaning

#### Note: Dataset was downloaded on November 7, 2022

Original dataset: https://drive.google.com/file/d/151pywOysoyVTpCjNwprf2wBAEHxvT56M/view?usp=sharing

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Exploring the Original Dataset

In [2]:
incidents_data = pd.read_csv("Police_Dept_Incident_Reports_Original.csv")
incidents_data.head(5)

Incident Datetime Incident Date Incident Time  Incident Year  \
0  2021/07/25 12:00:00 AM    2021/07/25         00:00           2021   
1  2022/06/28 11:58:00 PM    2022/06/28         23:58           2022   
2  2022/03/11 10:30:00 AM    2022/03/11         10:30           2022   
3  2021/05/15 05:47:00 PM    2021/05/15         17:47           2021   
4  2022/06/28 05:22:00 PM    2022/06/28         17:22           2022   

  Incident Day of Week         Report Datetime        Row ID  Incident ID  \
0               Sunday  2021/07/25 01:41:00 PM  105718906372      1057189   
1              Tuesday  2022/06/28 11:58:00 PM  116554371012      1165543   
2               Friday  2022/03/11 08:03:00 PM  113048071000      1130480   
3             Saturday  2021/05/15 05:47:00 PM  103051807043      1030518   
4              Tuesday  2022/06/28 05:22:00 PM  116535107041      1165351   

   Incident Number  CAD Number  ... Longitude Point Neighborhoods  \
0        216105573         NaN  ...       NaN   NaN           NaN   
1        220264913         NaN  ...       NaN   NaN           NaN   
2        226040232         NaN  ...       NaN   NaN           NaN   
3        210183345         NaN  ...       NaN   NaN           NaN   
4        220361741         NaN  ...       NaN   NaN           NaN   

   ESNCAG - Boundary File  \
0                     NaN   
1                     NaN   
2                     NaN   
3                     NaN   
4                     NaN   

  Central Market/Tenderloin Boundary Polygon - Updated  \
0                                                NaN     
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

  Civic Center Harm Reduction Project Boundary HSOC Zones as of 2018-06-05  \
0                                          NaN                         NaN   
1                                          NaN                         NaN   
2                                          NaN                         NaN   
3                                          NaN                         NaN   
4                                          NaN                         NaN   

  Invest In Neighborhoods (IIN) Areas Current Supervisor Districts  \
0                                 NaN                          NaN   
1                                 NaN                          NaN   
2                                 NaN                          NaN   
3                                 NaN                          NaN   
4                                 NaN                          NaN   

   Current Police Districts  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  

[5 rows x 34 columns]

In [3]:
incidents_data.shape

(658728, 34)

- The original dataset has 658,728 rows and 34 columns.

### Data Types

In [4]:
incidents_data.dtypes

Incident Datetime                                        object
Incident Date                                            object
Incident Time                                            object
Incident Year                                             int64
Incident Day of Week                                     object
Report Datetime                                          object
Row ID                                                    int64
Incident ID                                               int64
Incident Number                                           int64
CAD Number                                              float64
Report Type Code                                         object
Report Type Description                                  object
Filed Online                                             object
Incident Code                                             int64
Incident Category                                        object
Incident Subcategory                    

### Null Values

In [5]:
incidents_data.isnull().sum()

Incident Datetime                                            0
Incident Date                                                0
Incident Time                                                0
Incident Year                                                0
Incident Day of Week                                         0
Report Datetime                                              0
Row ID                                                       0
Incident ID                                                  0
Incident Number                                              0
CAD Number                                              147096
Report Type Code                                             0
Report Type Description                                      0
Filed Online                                            526011
Incident Code                                                0
Incident Category                                          556
Incident Subcategory                                   

- There are many null values in some of the columns. We will probably remove these columns, as we don't really need them to answer our exploratory questions.

## Filtering the Original Dataset

### We only want the rows where the *Incident Subcategory* is "Larceny - From Vehicle":

In [6]:
car_thefts = incidents_data[incidents_data['Incident Subcategory'] == 'Larceny - From Vehicle']
car_thefts.head()

Incident Datetime Incident Date Incident Time  Incident Year  \
6   2022/06/28 02:00:00 PM    2022/06/28         14:00           2022   
12  2022/06/23 12:00:00 AM    2022/06/23         00:00           2022   
14  2021/09/28 08:00:00 PM    2021/09/28         20:00           2021   
19  2021/11/20 08:30:00 PM    2021/11/20         20:30           2021   
22  2021/11/14 08:30:00 PM    2021/11/14         20:30           2021   

   Incident Day of Week         Report Datetime        Row ID  Incident ID  \
6               Tuesday  2022/06/28 03:09:00 PM  116546206244      1165462   
12             Thursday  2022/06/28 12:03:00 AM  116550406224      1165504   
14              Tuesday  2021/09/30 09:41:00 AM  107641306244      1076413   
19             Saturday  2021/11/21 02:34:00 PM  109435606224      1094356   
22               Sunday  2021/11/14 10:21:00 PM  109431206224      1094312   

    Incident Number  CAD Number  ... Longitude Point Neighborhoods  \
6         226109026         NaN  ...       NaN   NaN           NaN   
12        226108777         NaN  ...       NaN   NaN           NaN   
14        216140660         NaN  ...       NaN   NaN           NaN   
19        216178376         NaN  ...       NaN   NaN           NaN   
22        216178837         NaN  ...       NaN   NaN           NaN   

    ESNCAG - Boundary File  \
6                      NaN   
12                     NaN   
14                     NaN   
19                     NaN   
22                     NaN   

   Central Market/Tenderloin Boundary Polygon - Updated  \
6                                                 NaN     
12                                                NaN     
14                                                NaN     
19                                                NaN     
22                                                NaN     

   Civic Center Harm Reduction Project Boundary HSOC Zones as of 2018-06-05  \
6                                           NaN                         NaN   
12                                          NaN                         NaN   
14                                          NaN                         NaN   
19                                          NaN                         NaN   
22                                          NaN                         NaN   

   Invest In Neighborhoods (IIN) Areas Current Supervisor Districts  \
6                                  NaN                          NaN   
12                                 NaN                          NaN   
14                                 NaN                          NaN   
19                                 NaN                          NaN   
22                                 NaN                          NaN   

    Current Police Districts  
6                        NaN  
12                       NaN  
14                       NaN  
19                       NaN  
22                       NaN  

[5 rows x 34 columns]

In [7]:
car_thefts.shape

(113814, 34)

### We also only want the rows where the *Police District* is not "Out of SF":

In [8]:
sf_car_thefts = car_thefts[car_thefts['Police District'] != 'Out of SF']
sf_car_thefts.head()

Incident Datetime Incident Date Incident Time  Incident Year  \
6   2022/06/28 02:00:00 PM    2022/06/28         14:00           2022   
12  2022/06/23 12:00:00 AM    2022/06/23         00:00           2022   
14  2021/09/28 08:00:00 PM    2021/09/28         20:00           2021   
19  2021/11/20 08:30:00 PM    2021/11/20         20:30           2021   
22  2021/11/14 08:30:00 PM    2021/11/14         20:30           2021   

   Incident Day of Week         Report Datetime        Row ID  Incident ID  \
6               Tuesday  2022/06/28 03:09:00 PM  116546206244      1165462   
12             Thursday  2022/06/28 12:03:00 AM  116550406224      1165504   
14              Tuesday  2021/09/30 09:41:00 AM  107641306244      1076413   
19             Saturday  2021/11/21 02:34:00 PM  109435606224      1094356   
22               Sunday  2021/11/14 10:21:00 PM  109431206224      1094312   

    Incident Number  CAD Number  ... Longitude Point Neighborhoods  \
6         226109026         NaN  ...       NaN   NaN           NaN   
12        226108777         NaN  ...       NaN   NaN           NaN   
14        216140660         NaN  ...       NaN   NaN           NaN   
19        216178376         NaN  ...       NaN   NaN           NaN   
22        216178837         NaN  ...       NaN   NaN           NaN   

    ESNCAG - Boundary File  \
6                      NaN   
12                     NaN   
14                     NaN   
19                     NaN   
22                     NaN   

   Central Market/Tenderloin Boundary Polygon - Updated  \
6                                                 NaN     
12                                                NaN     
14                                                NaN     
19                                                NaN     
22                                                NaN     

   Civic Center Harm Reduction Project Boundary HSOC Zones as of 2018-06-05  \
6                                           NaN                         NaN   
12                                          NaN                         NaN   
14                                          NaN                         NaN   
19                                          NaN                         NaN   
22                                          NaN                         NaN   

   Invest In Neighborhoods (IIN) Areas Current Supervisor Districts  \
6                                  NaN                          NaN   
12                                 NaN                          NaN   
14                                 NaN                          NaN   
19                                 NaN                          NaN   
22                                 NaN                          NaN   

    Current Police Districts  
6                        NaN  
12                       NaN  
14                       NaN  
19                       NaN  
22                       NaN  

[5 rows x 34 columns]

In [9]:
sf_car_thefts.shape

(113436, 34)

## Duplicated *Incident IDs*

This is just to check whether or not there are duplicated *Incident IDs*.

In [10]:
print("Number of unique incident IDs:", sf_car_thefts['Incident ID'].nunique())

Number of unique incident IDs: 113400


- There are 113,436 rows in the SF car thefts dataset but 113,400 unique incident IDs, which likely means that there are multiple entries for the same incident ID.

- The data documentation does state that multiple reports of the same incident will be different rows but have the same incident ID. Updated files will be new rows but will share the same ID as well.

- We will remove these duplicates and keep only the initial incident filing, as it wouldn’t make too much sense to include the same incident multiple times in our analysis.

In [11]:
sf_car_thefts = sf_car_thefts.drop_duplicates(subset = 'Incident ID', keep = 'first')
sf_car_thefts

Incident Datetime Incident Date Incident Time  Incident Year  \
6       2022/06/28 02:00:00 PM    2022/06/28         14:00           2022   
12      2022/06/23 12:00:00 AM    2022/06/23         00:00           2022   
14      2021/09/28 08:00:00 PM    2021/09/28         20:00           2021   
19      2021/11/20 08:30:00 PM    2021/11/20         20:30           2021   
22      2021/11/14 08:30:00 PM    2021/11/14         20:30           2021   
...                        ...           ...           ...            ...   
658700  2022/10/30 12:00:00 PM    2022/10/30         12:00           2022   
658704  2022/11/03 04:45:00 PM    2022/11/03         16:45           2022   
658709  2022/11/04 06:00:00 AM    2022/11/04         06:00           2022   
658710  2022/11/01 06:34:00 AM    2022/11/01         06:34           2022   
658720  2022/10/29 11:30:00 AM    2022/10/29         11:30           2022   

       Incident Day of Week         Report Datetime        Row ID  \
6                   Tuesday  2022/06/28 03:09:00 PM  116546206244   
12                 Thursday  2022/06/28 12:03:00 AM  116550406224   
14                  Tuesday  2021/09/30 09:41:00 AM  107641306244   
19                 Saturday  2021/11/21 02:34:00 PM  109435606224   
22                   Sunday  2021/11/14 10:21:00 PM  109431206224   
...                     ...                     ...           ...   
658700               Sunday  2022/11/04 12:49:00 PM  121155106224   
658704             Thursday  2022/11/05 04:52:00 PM  121172306243   
658709               Friday  2022/11/04 08:40:00 PM  121149806224   
658710              Tuesday  2022/11/04 01:46:00 PM  121159906244   
658720             Saturday  2022/10/29 12:00:00 PM  121156006224   

        Incident ID  Incident Number   CAD Number  ...   Longitude  \
6           1165462        226109026          NaN  ...         NaN   
12          1165504        226108777          NaN  ...         NaN   
14          1076413        216140660          NaN  ...         NaN   
19          1094356        216178376          NaN  ...         NaN   
22          1094312        216178837          NaN  ...         NaN   
...             ...              ...          ...  ...         ...   
658700      1211551        226208355          NaN  ... -122.438206   
658704      1211723        220762597  223092177.0  ... -122.450376   
658709      1211498        220760858  223082577.0  ... -122.387176   
658710      1211599        220753110          NaN  ... -122.511295   
658720      1211560        226208145          NaN  ... -122.394060   

                                                 Point Neighborhoods  \
6                                                  NaN           NaN   
12                                                 NaN           NaN   
14                                                 NaN           NaN   
19                                                 NaN           NaN   
22                                                 NaN           NaN   
...                                                ...           ...   
658700  POINT (-122.43820638374945 37.785459340995516)         103.0   
658704  POINT (-122.45037621389666 37.748836616473476)          48.0   
658709   POINT (-122.38717561710232 37.73566160727889)          86.0   
658710   POINT (-122.51129492624534 37.77507596005672)           8.0   
658720  POINT (-122.39405962490252 37.792537216359484)         108.0   

        ESNCAG - Boundary File  \
6                          NaN   
12                         NaN   
14                         NaN   
19                         NaN   
22                         NaN   
...                        ...   
658700                     NaN   
658704                     NaN   
658709                     NaN   
658710                     NaN   
658720                     NaN   

       Central Market/Tenderloin Boundary Polygon - Updated  \
6                                                     NaN     
12                

## Changing *Incident Date* Column to YYYY-MM-DD Format 

In [12]:
pd.options.mode.chained_assignment = None
sf_car_thefts['Incident Date'] = pd.to_datetime(sf_car_thefts['Incident Date'])
sf_car_thefts

Incident Datetime Incident Date Incident Time  Incident Year  \
6       2022/06/28 02:00:00 PM    2022-06-28         14:00           2022   
12      2022/06/23 12:00:00 AM    2022-06-23         00:00           2022   
14      2021/09/28 08:00:00 PM    2021-09-28         20:00           2021   
19      2021/11/20 08:30:00 PM    2021-11-20         20:30           2021   
22      2021/11/14 08:30:00 PM    2021-11-14         20:30           2021   
...                        ...           ...           ...            ...   
658700  2022/10/30 12:00:00 PM    2022-10-30         12:00           2022   
658704  2022/11/03 04:45:00 PM    2022-11-03         16:45           2022   
658709  2022/11/04 06:00:00 AM    2022-11-04         06:00           2022   
658710  2022/11/01 06:34:00 AM    2022-11-01         06:34           2022   
658720  2022/10/29 11:30:00 AM    2022-10-29         11:30           2022   

       Incident Day of Week         Report Datetime        Row ID  \
6                   Tuesday  2022/06/28 03:09:00 PM  116546206244   
12                 Thursday  2022/06/28 12:03:00 AM  116550406224   
14                  Tuesday  2021/09/30 09:41:00 AM  107641306244   
19                 Saturday  2021/11/21 02:34:00 PM  109435606224   
22                   Sunday  2021/11/14 10:21:00 PM  109431206224   
...                     ...                     ...           ...   
658700               Sunday  2022/11/04 12:49:00 PM  121155106224   
658704             Thursday  2022/11/05 04:52:00 PM  121172306243   
658709               Friday  2022/11/04 08:40:00 PM  121149806224   
658710              Tuesday  2022/11/04 01:46:00 PM  121159906244   
658720             Saturday  2022/10/29 12:00:00 PM  121156006224   

        Incident ID  Incident Number   CAD Number  ...   Longitude  \
6           1165462        226109026          NaN  ...         NaN   
12          1165504        226108777          NaN  ...         NaN   
14          1076413        216140660          NaN  ...         NaN   
19          1094356        216178376          NaN  ...         NaN   
22          1094312        216178837          NaN  ...         NaN   
...             ...              ...          ...  ...         ...   
658700      1211551        226208355          NaN  ... -122.438206   
658704      1211723        220762597  223092177.0  ... -122.450376   
658709      1211498        220760858  223082577.0  ... -122.387176   
658710      1211599        220753110          NaN  ... -122.511295   
658720      1211560        226208145          NaN  ... -122.394060   

                                                 Point Neighborhoods  \
6                                                  NaN           NaN   
12                                                 NaN           NaN   
14                                                 NaN           NaN   
19                                                 NaN           NaN   
22                                                 NaN           NaN   
...                                                ...           ...   
658700  POINT (-122.43820638374945 37.785459340995516)         103.0   
658704  POINT (-122.45037621389666 37.748836616473476)          48.0   
658709   POINT (-122.38717561710232 37.73566160727889)          86.0   
658710   POINT (-122.51129492624534 37.77507596005672)           8.0   
658720  POINT (-122.39405962490252 37.792537216359484)         108.0   

        ESNCAG - Boundary File  \
6                          NaN   
12                         NaN   
14                         NaN   
19                         NaN   
22                         NaN   
...                        ...   
658700                     NaN   
658704                     NaN   
658709                     NaN   
658710                     NaN   
658720                     NaN   

       Central Market/Tenderloin Boundary Polygon - Updated  \
6                                                     NaN     
12                

## Adding Columns

- We need the month of the incident for our data analysis section, so we will extract the month as a separate column called `Incident Month`.

In [13]:
sf_car_thefts['Incident Month'] = sf_car_thefts['Incident Date'].dt.month

## Dropping Columns

- For now, we will only keep 17 columns out of the original dataset: `Incident Datetime`, `Incident Date`, `Incident Time`, `Incident Year`, `Incident Day of Week`, `Report Datetime`, `Row ID`, `Incident ID`, `Incident Number`, `Report Type Code`, `Report Type Description`, `Incident Code`, `Incident Description`, `Resolution`, `Incident Category`, `Incident Subcategory`, and `Police District`.

- We might not use every column of the 17 being kept. I just wanted to be safe, in case there's a need for them further down the line.

- The remaining 17 columns either have too many null values, will not help us in answering our questions, or both - so we will drop them.

In [14]:
sf_car_thefts = sf_car_thefts.drop(columns = ['CAD Number', 'Filed Online', 'Intersection', 'CNN',
                                        'Analysis Neighborhood', 'Supervisor District', 'Latitude',
                                        'Longitude', 'Point', 'Neighborhoods', 'ESNCAG - Boundary File',
                                        'Central Market/Tenderloin Boundary Polygon - Updated',
                                        'Civic Center Harm Reduction Project Boundary', 'HSOC Zones as of 2018-06-05',
                                        'Invest In Neighborhoods (IIN) Areas', 'Current Supervisor Districts',
                                        'Current Police Districts'])
sf_car_thefts

Incident Datetime Incident Date Incident Time  Incident Year  \
6       2022/06/28 02:00:00 PM    2022-06-28         14:00           2022   
12      2022/06/23 12:00:00 AM    2022-06-23         00:00           2022   
14      2021/09/28 08:00:00 PM    2021-09-28         20:00           2021   
19      2021/11/20 08:30:00 PM    2021-11-20         20:30           2021   
22      2021/11/14 08:30:00 PM    2021-11-14         20:30           2021   
...                        ...           ...           ...            ...   
658700  2022/10/30 12:00:00 PM    2022-10-30         12:00           2022   
658704  2022/11/03 04:45:00 PM    2022-11-03         16:45           2022   
658709  2022/11/04 06:00:00 AM    2022-11-04         06:00           2022   
658710  2022/11/01 06:34:00 AM    2022-11-01         06:34           2022   
658720  2022/10/29 11:30:00 AM    2022-10-29         11:30           2022   

       Incident Day of Week         Report Datetime        Row ID  \
6                   Tuesday  2022/06/28 03:09:00 PM  116546206244   
12                 Thursday  2022/06/28 12:03:00 AM  116550406224   
14                  Tuesday  2021/09/30 09:41:00 AM  107641306244   
19                 Saturday  2021/11/21 02:34:00 PM  109435606224   
22                   Sunday  2021/11/14 10:21:00 PM  109431206224   
...                     ...                     ...           ...   
658700               Sunday  2022/11/04 12:49:00 PM  121155106224   
658704             Thursday  2022/11/05 04:52:00 PM  121172306243   
658709               Friday  2022/11/04 08:40:00 PM  121149806224   
658710              Tuesday  2022/11/04 01:46:00 PM  121159906244   
658720             Saturday  2022/10/29 12:00:00 PM  121156006224   

        Incident ID  Incident Number Report Type Code Report Type Description  \
6           1165462        226109026               II        Coplogic Initial   
12          1165504        226108777               II        Coplogic Initial   
14          1076413        216140660               II        Coplogic Initial   
19          1094356        216178376               II        Coplogic Initial   
22          1094312        216178837               II        Coplogic Initial   
...             ...              ...              ...                     ...   
658700      1211551        226208355               II        Coplogic Initial   
658704      1211723        220762597               II                 Initial   
658709      1211498        220760858               II                 Initial   
658710      1211599        220753110               IS     Coplogic Supplement   
658720      1211560        226208145               II        Coplogic Initial   

        Incident Code Incident Category    Incident Subcategory  \
6                6244     Larceny Theft  Larceny - From Vehicle   
12               6224     Larceny Theft  Larceny - From Vehicle   
14               6244     Larceny Theft  Larceny - From Vehicle   
19               6224     Larceny Theft  Larceny - From Vehicle   
22               6224     Larceny Theft  Larceny - From Vehicle   
...               ...               ...                     ...   
658700           6224     Larceny Theft  Larceny - From Vehicle   
658704           6243     Larceny Theft  Larceny - From Vehicle   
658709           6224     Larceny Theft  Larceny - From Vehicle   
658710           6244     Larceny Theft  Larceny - From Vehicle   
658720           6224     Larceny Theft  Larceny - From Vehicle   

                         Incident Description      Resolution Police District  \
6           Theft, From Locked Vehicle, >$950  Open or Active        Richmond   
12        Theft, From Unlocked Vehicle, >$950  Open or Active         Mission   
14          Theft, From Locked Vehicle, >$950  Open or Active         Central   
19        Theft, From Unlocked Vehicle, >$950  Open or Active         Mission   
22        Theft, From Unlocked Vehicle, >$950  Open or Active         Centra

## Final Checks: Data Types and Null Values

In [15]:
sf_car_thefts.dtypes

Incident Datetime                  object
Incident Date              datetime64[ns]
Incident Time                      object
Incident Year                       int64
Incident Day of Week               object
Report Datetime                    object
Row ID                              int64
Incident ID                         int64
Incident Number                     int64
Report Type Code                   object
Report Type Description            object
Incident Code                       int64
Incident Category                  object
Incident Subcategory               object
Incident Description               object
Resolution                         object
Police District                    object
Incident Month                      int64
dtype: object

In [16]:
sf_car_thefts.isnull().sum()

Incident Datetime          0
Incident Date              0
Incident Time              0
Incident Year              0
Incident Day of Week       0
Report Datetime            0
Row ID                     0
Incident ID                0
Incident Number            0
Report Type Code           0
Report Type Description    0
Incident Code              0
Incident Category          0
Incident Subcategory       0
Incident Description       0
Resolution                 0
Police District            0
Incident Month             0
dtype: int64

## Saving as CSV File

In [17]:
sf_car_thefts.to_csv("sf_car_thefts.csv")